File.macho create (#1097)
elastic/ecs: rfcs/text/0000-create-file-macho.md

# 0000: Create the Mach-O sub-field of the File fieldset

> - Stage: **1 (proposal)**

Reviewer: ```suggestion
- Stage: **1 (draft)**
```

Same change to the stage naming, again 😂 . I think this is the last one! 🤞

- Date: **TBD**

> Create the Mach Object (Mach-O) sub-field, of the `file` top-level fieldset. This document metadata can be used for malware research, as well as coding and other application development efforts.

Reviewer: Can we add the intent to nest `macho` underneath `process`, as well?

> Create the Mach Object (Mach-O) sub-field, of the `file` top-level fieldset. This document metadata can be used for malware research, as well as coding and other application development efforts.

Author: ```suggestion
Create the Mach Object (Mach-O) sub-field, of the `file` or `process` top-level fieldsets. This document metadata can be used for malware research, as well as coding and other application development efforts.
```

## Fields

**Stage 0**

This RFC is to create the Mach-O sub-field within the `file.` fieldset. This will include 35 sub-fields.

|   Name                                     |   Type     |   Description                                                               |
|--------------------------------------------|------------|-----------------------------------------------------------------------------|
|   macho.cpu                          |   object   |   CPU information for the file.                                             |

> |   macho.cpu.architecture             |   keyword  |   CPU architecture target for the file.                                     |

Reviewer: @andrewstucki I did some cross-checking between the proposed field list here and your work in https://github.com/elastic/beats/pull/24195. I pulled those mappings for `macho` into [a gist](https://gist.github.com/ebeahan/41507b47bcadff43f866350634ffa9ae) for easier reference.

100% alignment isn't necessary at all at this point. However, there are enough differences between that approach and what's captured here I think it's worth raising.

@peasead @dcode @devonakerr

Reviewer 2: I have since updated this PR with a proposed layout that aligns sections and segments with ELF and PE (which doesn't have segments).

Reviewer: Thanks, @dcode. 

There's still some large differences between the implementation in elastic/beats#24195 and what's proposed here.

Again, not something we need to completely resolve now. At a minimum, let's capture the difference in approaches as a `Concern` to make sure we revisit the topic in the next stage's PR.

Reviewer 2: @andrewstucki the strategy I'm suggesting here is splitting out fat binary components that can be unified by the same file metadata (namely hashes or other source unique id's). That will allow analyzing all binaries by their sections, segments, symbols, etc in a seamless way, regardless of platform or if their multi-platform. Do you think that can work for your implementation?

Author: Just pining to see if there was a resolution here.

Reviewer 3: Just as a use-case example: let's say we're triggering an alert on a malicious binary that has been compiled for both the new m1 as well as older, intel-based chipsets and we want to add in object-level data about the file. Would we emit two separate alerts, one for each architecture? Seems kind of odd.

|   macho.header.commands.type        |   keyword  |   Type of the load commands for the Mach-O header.                          |
|   macho.header.magic                |   keyword  |   Magic field of the Mach-O header.                                         |
|   macho.header.flags                |   keyword  |   Flags set in the Mach-O header.                                           |
|   macho.segments                     |   flattened   |   Segment information for the file.                                         |
|   macho.segments.vmaddr              |   keyword  |   Memory address of this segment.                                           |
|   macho.segments.name                |   keyword  |   Name of this segment.                                                     |
|   macho.segments.vmsize              |   keyword  |   Memory size of this segment.                                              |
|   macho.segments.fileoff             |   keyword  |   File offset of this segment.                                              |
|   macho.segments.filesize            |   keyword  |   Amount of memory to map from the file.                                    |

> |   macho.segments.sections            |   object   |   Section information for the segment of the file.                          |

Reviewer 3: this is `flattened` in the field document below.

Author: Fix't

|   macho.segments.sections.name       |   keyword  |   Section name for the segment of the file.                                 |

> |   macho.segments.sections.type       |   keyword  |   Section type for the segment of the file.                                 |

Reviewer: Similar to what we discussed https://github.com/elastic/ecs/pull/1077#discussion_r572274291, `flattened` cannot have explicit fields mapped underneath.

If the data is expected to be shaped to an expected object format, we'll have to capture this another way.

Author: Do you mean that `macho.segments.sections` needs to be `nested`?

Reviewer: No, I mean if we use `flattened` and are expecting specific keys to be present under the `flattened` parent field, we'll need someone to capture this expectation in ECS.

Author: It does appear that there are specific keys under `macho.segments.sections.*`. @dcode is this something you can help with?

Here's a reference that could help: https://github.com/aidansteele/osx-abi-macho-file-format-reference#section under `flags`.

Reviewer 2: The fields under `macho.segments.sections[*]` is well-formed and consists of `flags` that denote how the memory is handled by the loader, the name of the section, the section name, and the type of section that it is. In ELF and PE, we put section offset information in `file.*.sections[*]`. This data is missing from the current rendering of the schema. Both the section and the segment have offset data. While in macho, sections always exist within a specific segment, they are actually two different views into the executable and are processed differently at runtime.

Reviewer: I suggest we go with using `nested` for `macho.segments.sections.*`. Any objection?

Reviewer 2: I agree with nested, but I think segments and sections should ideally match the usage as in ELF. Namely, we pull out sections to `file.macho.sections` and segments stay at `file.macho.segments`. The data source above nests them and it isn't wrong, but they are two different views into the binary. Sections also have physical offset data.

I'm going to resolve this conversation and suggest those changes (I can't seem to do it with this thread in the middle of it).

|   macho.segments.sections.type       |   keyword  |   Section type for the segment of the file.                                 |

> |   macho.segments.offset      |   keyword  |   Offset of the segment.                                             |

Reviewer 3: this doesn't match the type in the field definition below

Author: Fixed.

> |   macho.segments                     |   nested   |   Segment information for the file.                                         |
> |   macho.segments.vmaddr              |   keyword  |   Memory address of this segment.                                           |
> |   macho.segments.name                |   keyword  |   Name of this segment.                                                     |
> |   macho.segments.vmsize              |   keyword  |   Memory size of this segment.                                              |
> |   macho.segments.fileoff             |   keyword  |   File offset of this segment.                                              |
> |   macho.segments.filesize            |   keyword  |   Amount of memory to map from the file.                                    |
> |   macho.segments.sections            |   flattened   |   Section information for the segment of the file.                          |
> |   macho.segments.sections.flags      |   keyword  |   Section flags for the segment of the file.                                |
> |   macho.segments.sections.name       |   keyword  |   Section name for the segment of the file.                                 |
> |   macho.segments.sections.type       |   keyword  |   Section type for the segment of the file.                                 |
> |   macho.segments.offset      |   long  |   Offset of the segment.                                             |
> |   macho.segments.size     |   keyword  |   Segment limit size.                                            |
> |   macho.segments.flags     |   keyword  |   Segment flags.                                                 |

Reviewer 2: This suggestion matches the data interface to ELF, using consistent naming (physical offset vs file offset for example), aligns segments and sections information, and adds `entropy` and  `chi2` values to `macho.sections`.
```suggestion
|   macho.segments                 |   nested   |   Segment information for the file.                   |
|   macho.segments.name            |   keyword  |   Name of this segment.                               |
|   macho.segments.physical_offset |   long     |   File offset of this segment.                        |
|   macho.segments.physical_size   |   keyword  |   Amount of memory to map from the file.              |
|   macho.segments.virtual_address |   keyword  |   Memory address of this segment.                     |
|   macho.segments.virtual_size    |   keyword  |   Memory size of this segment.                        |
|   macho.segments.sections        |   keyword  |   Section names contained in this segment.            |
|   macho.sections                 |   nested   |   Section information for the segment of the file.    |
|   macho.sections.name            |   keyword  |   Section name for the segment of the file.           |
|   macho.sections.flags           |   keyword  |   Section flags for the segment of the file.          |
|   macho.sections.type            |   keyword  |   Section type for the segment of the file.           |
|   macho.sections.physical_offset |   long     |   Section List offset.                                |
|   macho.sections.physical_size   |   long     |   Section List physical size.                         |
|   macho.sections.virtual_address |   long     |   Section List virtual address.                       |
|   macho.sections.virtual_size    |   long     |   Section List virtual size.                          |
|   macho.sections.entropy         |   float    |   Shannon entropy calculation from the section.       |
|   macho.sections.chi2            |   float    |   Chi-square probability distribution of the section. |
```

> |   macho.segment size     |   keyword  |   Segment limit size.                                            |

Reviewer 3: can you clean this up? think it should be `macho.segments.size` and `segments` should be moved to `nested` both here and in the field definitions.

Author: Done

> |   macho.segment.flags     |   keyword  |   Segment flags.                                                 |

Reviewer 3: `macho.segments.flags`?

Author: Fixed.

As another example, tracking file metadata for specific families is useful in predicting new campaigns if we see similar file metadata being used for new samples. [Example](https://www.bleepingcomputer.com/news/security/maze-ransomware-is-shutting-down-its-cybercrime-operation/), the Maze ransomware family shutting down and re-purposing as Egregor (this is for Windows malware, but the concept is the same).

## Source data

**Stage 1**

This type of data can be provided by logs from VirusTotal, Reversing Labs, Lockheed Martin's LAIKABOSS, Emerson's File Scanning Framework, Target's Strelka, or other file/malware analysis platforms.

* [VirusTotal API](https://developers.virustotal.com/v3.0/reference)
* [Emerson FSF](https://github.com/EmersonElectricCo/fsf)
* [Target Strelka](https://github.com/target/strelka)
* [Lockheed Martin LAIKABOSS](https://github.com/lmco/laikaboss)

> 

Reviewer 2: ```suggestion
* [LIEF Analysis Library](https://lief.quarkslab.com/doc/latest/api/python/macho.html)

```

-->

<!--
Stage 3: Add more real world example source documents so we have at least 2 total, but ideally 3. Format as described in stage 2.
-->

## Scope of impact

**Stage 2**

There should be no breaking changes, depreciation strategies, or significant refactoring as this is creating a sub-field for the existing `file.` fieldset.

While likely not a large-scale ECS project, there would be documentation updates needed to explain the new fields.

<!--
Stage 2: Identifies scope of impact of changes. Are breaking changes required? Should deprecation strategies be adopted? Will significant refactoring be involved? Break the impact down into:
 * Ingestion mechanisms (e.g. beats/logstash)
 * Usage mechanisms (e.g. Kibana applications, detections)
 * ECS project (e.g. docs, tooling)
The goal here is to research and understand the impact of these changes on users in the community and development teams across Elastic. 2-5 sentences each.
-->

## Concerns

> 

Reviewer: Any known concerns we should capture at this point?

-->

<!--
Stage 3: Document resolutions for all existing concerns. Any new concerns should be documented along with their resolution. The goal here is to eliminate the risk of churn and instability by resolving outstanding concerns.
-->

<!--
Stage 4: Document any new concerns and their resolution. The goal here is to eliminate risk of churn and instability by ensuring all concerns have been addressed.
-->

## Real-world implementations

<!--
Stage 4: Identify at least one real-world, production-ready implementation that uses these updated field definitions. An example of this might be a GA feature in an Elastic application in Kibana.
-->

## People

The following are the people that consulted on the contents of this RFC.

* @peasead | author
* @devonakerr | sponsor
* @dcode, @peasead | subject matter expert

## References

<!-- Insert any links appropriate to this RFC in this section. -->

### RFC Pull Requests

<!-- An RFC should link to the PRs for each of it stage advancements. -->

> * Stage 0: https://github.com/elastic/ecs/pull/1097

Reviewer: ```suggestion
* Stage 1: https://github.com/elastic/ecs/pull/1097
```
